relay-server/src/utils/envelope_context.rs

```diff
@@ -91,23 +99,41 @@ impl EnvelopeContext {
     /// As opposed to [`new`](Self::new), this does not require a queue permit. This makes it
     /// suitable for unit testing internals of the processing pipeline.
     #[cfg(test)]
-    pub fn standalone(envelope: &Envelope) -> Self {
+    pub fn standalone(envelope: Box<Envelope>) -> Self {
         Self::new_internal(envelope, None)
     }
 
     /// Computes an envelope context from the given envelope and binds it to the processing queue.
     ///
     /// To provide additional scoping, use [`EnvelopeContext::scope`].
-    pub fn new(envelope: &Envelope, slot: SemaphorePermit) -> Self {
+    pub fn new(envelope: Box<Envelope>, slot: SemaphorePermit) -> Self {
         Self::new_internal(envelope, Some(slot))
     }
 
+    /// Returns a reference to the contained [`Envelope`].
+    pub fn envelope(&self) -> &Envelope {
+        self.envelope.as_ref()
+    }
+
+    /// Returns a mutable reference to the contained [`Envelope`].
+    pub fn envelope_mut(&mut self) -> &mut Envelope {
+        self.envelope.as_mut()
+    }
+
+    /// Take the envelope out of the context and replace it with a dummy.
+    ///
+    /// Note that after taking out the envelope, the envelope summary is incorrect.
+    pub(crate) fn take_envelope(&mut self) -> Box<Envelope> {
+        let dummy_envelope = Envelope::parse_bytes(Bytes::from("{}")).unwrap();
+        std::mem::replace(&mut self.envelope, dummy_envelope)
+    }
```

Comment:
This method should not exist because it leaves the `EnvelopeContext` in an inconsistent state. It is only used in `handle_submit` at the end of the envelope's life time.